Support undo/redo for control points in the Editor

smoogipoo opened the pull request:
Works towards resolving https://github.com/ppy/osu/issues/10314. Now handles control points.

... In the interim until we have binary object comparisons.

smoogipoo committed:
Support undo/redo for control points

smoogipoo committed:
Work around performance issues

turbedi commented on osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:
```diff
@@ -32,61 +34,99 @@ public void Patch(byte[] currentState, byte[] newState)
-            int oldHitObjectsIndex = Array.IndexOf(result.PiecesOld, "[HitObjects]");
-            int newHitObjectsIndex = Array.IndexOf(result.PiecesNew, "[HitObjects]");
+            int oldSectionStartIndex = Array.IndexOf(result.PiecesOld, $"[{section}]");
+            int oldSectionEndIndex = Array.FindIndex(result.PiecesOld, oldSectionStartIndex + 1, s => s.StartsWith(@"[", StringComparison.Ordinal));
```
```suggestion
            int oldSectionEndIndex = Array.FindIndex(result.PiecesOld, oldSectionStartIndex + 1, s => s.StartsWith('['));
```

turbedi commented on osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:
```diff
@@ -32,61 +34,99 @@ public void Patch(byte[] currentState, byte[] newState)
+                oldSectionEndIndex = result.PiecesOld.Length;
+
+            int newSectionStartIndex = Array.IndexOf(result.PiecesNew, $"[{section}]");
+            int newSectionEndIndex = Array.FindIndex(result.PiecesNew, newSectionStartIndex + 1, s => s.StartsWith(@"[", StringComparison.Ordinal));
```
```suggestion
            int newSectionEndIndex = Array.FindIndex(result.PiecesNew, newSectionStartIndex + 1, s => s.StartsWith('['));
```

smoogipoo commented:
It doesn't seem to be perfect in all cases:
- Changing scroll speed doesn't revert in osu! mode. It _does_ revert in e.g. mania, where it actually has an effect.
- Changing slider velocity in the timeline doesn't revert.

These are issues elsewhere and relate to the way `DifficultyControlPoints` are extracted. It's a little bit of a mess, and this is a best effort.

smoogipoo committed:
Compare by `char`

Co-authored-by: Berkan Diler <berkan.diler1@ingka.ikea.com>

peppy commented:
Probably fine / much better than nothing at all. Not sure the scroll speed setting should even be visible there.

smoogipoo commented:
Sorry for the force push, I changed no earlier code, only removed a new commit that tried to fix `DifficultyControlPoints` not updating on `HitObject`s, then realised it caused damage on the actual timing screen :( Going to require more thought as previously said.

bdach commented:
Are the assertion failures from the earlier runs here ([example 1](https://github.com/ppy/osu/runs/6856350601?check_suite_focus=true), [example 2](https://github.com/ppy/osu/runs/6856481102?check_suite_focus=true)) related to the removed commit? As in, is this in a reviewable state right now?

smoogipoo commented:
Nah, I'm currently fixing the assertion failures.

smoogipoo committed:
Fix incorrect fallback value

smoogipoo committed:
Fix assertion failures

smoogipoo committed:
Add explanatory comment

smoogipoo commented:
Should be reviewable now.

bdach commented:
Also not sure if related, but there's a weird bug with the bottom timeline display where control points are supposed to be shown:

https://user-images.githubusercontent.com/20418176/173310853-b1c1fcaf-7417-4fde-97ee-62789e820629.mp4

Appears that consecutive undoes hide and show the control point indicators... for some reason? Not sure if related, just bringing it up as something I noticed.

smoogipoo commented:
The timeline issue could be related to incrementally adding every control point rather than giving a list with every control point populated at once. But probably a separate issue from this branch...?

smoogipoo commented on osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:
```diff
@@ -32,61 +34,105 @@ public void Patch(byte[] currentState, byte[] newState)
-            }
-
-            editorBeatmap.EndChange();
+            // The expected usage of this returned list is to iterate from the start to the end of the list, such that
+            // these indices need to appear in reverse order for the usage to not have to deal with decrementing indices.
+            removedIndices.Reverse();
```
Sure, I was also on the edge about this.

smoogipoo committed:
Apply refactorings from review

bdach reviewed (dismissed):
Looks ok to me

peppy commented:
This doesn't seem to handle the case where a new type is added to an existing group, is that known / expected?

smoogipoo commented:
Probably because the type isn't doing anything? It's only undoing/redoing based on what actually gets written to the file, so those points are probably becoming redundant & removed during write?

peppy commented:
Hmm... that's... very awkward. Especially since it doesn't clear redo states. So you hit redo and suddenly you are in a completely broken looking state.

peppy commented:
Something has regressed with selected rows on the timing screen. After cloning a point, the newly cloned point is not selected visually (but seems to have the selection internally). I've checked that this works correctly on `master`

https://user-images.githubusercontent.com/191335/173552802-ab04af4e-a401-4be0-a81e-297b0437da7c.mp4

- Select point
- Move to new time
- Clone

Observed row not selected.

smoogipoo commented:
Admittedly I didn't thoroughly test redoing so I didn't notice it breaking it, though your video doesn't show it.

peppy commented:
I had a video which has me redoing, but I didn't know what was going on and didn't think it would transfer well to a video 😅

smoogipoo committed:
Merge branch 'master' into editor-controlpoint-undo-redo

smoogipoo committed:
Implement IEquatable on ControlPoint

smoogipoo committed:
Fix selection by directly comparing control points

Previously, all control points would get replaced, which led to
performance issues that was worked around in this PR. By comparing
control points, we're able to get good performance without requiring the
workaround.

smoogipoo commented:
Please try this PR again and see if you can break it still. I think I might've resolved both the above issues by not replacing all control points.

There's still some weirdness (which I can't handle), that's best described with a video:
https://user-images.githubusercontent.com/1329837/174539856-fe30be08-2472-43f2-a87b-fb106dd0bbae.mp4

In the video, I'm doing much the same actions as https://github.com/ppy/osu/pull/18668#issuecomment-1154979283.
1. I add an effect point to the first timing point.
    - No undo/redo state created (redundant point).
2. I duplicate an effect point later on.
    - No undo/redo state created (redundant point).
3. I change the duplicated effect point to adjust kiai time.
    - Creates an undo state.
4. I press undo.
    - Both the duplicated effect point and the effect point added in (1) are removed.
5. I press redo.
    - The effect point added in (1) does not return.
    - The duplicated effect point is merged into the one following it (which also starts kiai), because the one following it has been deemed redundant.

peppy commented:
I can still get in a state where the row isn't selected. I'm not too concerned about certain changes remaining unsupported (like the scroll speed one in the video) but having the row non-selectable is quite breaking. Haven't looked at the code changes yet, but maybe a fix can be applied more locally to the table control rather than adding the overhead with equality etc?

https://user-images.githubusercontent.com/191335/174545301-cf83e153-5156-4991-a78b-a0c9364c4356.mp4

smoogipoo commented:
Looks like the breaking is reproed by:
1. Select timing point.
2. Adjust BPM.
3. Select any other point.
4. Undo.

> fix can be applied more locally to the table control rather than adding the overhead with equality etc

That's exactly what I was doing before, but it's even worse that way (breakage-wise) because performance starts to become a factor and you need to add _layers_ of workarounds to resolve it.

I prefer equality, but let's see if the above issue changes my view on things.

smoogipoo committed:
Fix group deselected when table is recreated

smoogipoo committed:
Properly annotate method to allow null

smoogipoo committed:
Convert == usages to ReferenceEquals

smoogipoo commented:
Above issue should be fixed. Though I understand there's quite a lot of files changed here, I still believe that control point equality is the way to go.

I can split the changes out if it makes it easier to review - these changes are MVP and I'd also like to annotate `ControlPointInfo` with #nullable but that's a bit more involved, so I could bundle this in a separate PR.

smoogipoo committed:
Compare by reference in ControlPoint.Equals()

smoogipoo committed:
Change to use ReferenceEquals

smoogipoo commented on osu.Game/Screens/Edit/Timing/ControlPointTable.cs:
```diff
@@ -44,6 +44,7 @@ public IEnumerable<ControlPointGroup> ControlGroups
                 {
                     BackgroundFlow.Add(new RowBackground(group)
                     {
+                        Selected = group.Equals(selectedGroup?.Value),
```
Nah, will change.

- I previously changed `Equals()` to compare by reference, but didn't include it in the end. I've changed it to include it now.
- Changed this to `ReferenceEquals`

smoogipoo committed:
Remove redundant diff processing

smoogipoo commented on osu.Game/Screens/Edit/LegacyEditorBeatmapPatcher.cs:
```diff
@@ -34,61 +37,112 @@ public void Patch(byte[] currentState, byte[] newState)
+
+        private void processTimingPoints(DiffResult result, Func<IBeatmap> getNewBeatmap)
+        {
+            findChangedIndices(result, LegacyDecoder<Beatmap>.Section.TimingPoints, out var removedIndices, out var addedIndices);
+
+            if (removedIndices.Count == 0 && addedIndices.Count == 0)
+                return;
```
Main concern was reverting hitobject changes on maps like https://osu.ppy.sh/beatmapsets/65994#osu/203906 that have many control points, but looks okay. The code below could also be changed to use hashset-based prechecks if this becomes an issue.

Removed.

peppy commented on osu.Game/Screens/Edit/Timing/ControlPointTable.cs:
```diff
@@ -44,6 +44,7 @@ public IEnumerable<ControlPointGroup> ControlGroups
                 {
                     BackgroundFlow.Add(new RowBackground(group)
                     {
+                        Selected = ReferenceEquals(group, selectedGroup?.Value),
```
The null check on `selectedGroup` here is causing code quality fail. I'd hope this isn't actually needed, but if it is then moving this to an `updateDisplay` or `Schedule`ing is a better option.

smoogipoo committed:
Split group selection to separate method

peppy reviewed (approved):
Seems pretty good
